nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "nav2_costmap_2d/costmap_subscriber.hpp"
+
+namespace nav2_costmap_2d
+{
+
+CostmapSubscriber::CostmapSubscriber(
+  rclcpp::Node::SharedPtr ros_node,
+  std::string & topic_name)
+: node_(ros_node),
+  topic_name_(topic_name),
+  costmap_received_(false),
+  costmap_(nullptr)
+{
+  costmap_sub_ = node_->create_subscription<nav2_msgs::msg::Costmap>(topic_name,
+      std::bind(&CostmapSubscriber::costmap_callback, this, std::placeholders::_1));
+}
+
+Costmap2D * CostmapSubscriber::getCostmap()
+{
+  if (costmap_ == nullptr) {
+    throw std::runtime_error("Costmap is not available");
+  }
+  return costmap_;
+}
+
+void CostmapSubscriber::toCostmap2D()
+{
+  if (!costmap_received_) {
+    costmap_ = new Costmap2D(
```

Comment:
Where to we `delete`? alternatively use `std::shared_ptr`

Reply:
Or unique_ptr if it is not shared.